### PR TITLE
Various small bug fixes in ffigen

### PR DESCRIPTION
**Pull request by liamappelbe**
Fixing some bugs I ran into while working on protocol implementation:

- `NSObject` is not actually the root of the ObjC class hierarchy. Instead there's an internal object type that's the true root. This is what `id` is actually supposed to map to. In the Dart wrapper objects, our root is `ObjCObjectBase`, so I've mapped `id` to this instead.
- Fixed most of the "Duplicate methods with different signatures" log spam.
  - Add method deps before checking for duplicates, to ensure that msgSend is non-null. Part of the signature check is checking whether the msgSends are identical between the methods, and if one is null then that's treated as a mismatch.
  - Some Apple APIs have methods that are supposed to be declared as returning `instancetype`, but instead return `id`, eg `NSInflectionRule.init`. In these cases we keep the `instancetype` version.
  - Found two other causes of this spam, but I'll leave those for later: #1136 and #1137
- Remove some nonexistent includes from pkgs/objective_c/src/foundation.h.

### CODE REVIEW

File: pkgs/ffigen/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 12.1.0-wip
+
+- Code-gen the ObjC `id` type to `ObjCObjectBase` rather than `NSObject`, since
```
Comment (dcharkes) [inline pkgs/ffigen/CHANGELOG.md:3]: Isn't this technically a breaking change?

**Review by dcharkes — APPROVED** [review]
> Fixed most of the "Duplicate methods with different signatures" log spam.

Please add a test for this, so that we don't regress this.

File: pkgs/ffigen/CHANGELOG.md (hunk 1 above)
Comment (liamappelbe) [reply pkgs/ffigen/CHANGELOG.md]: I think it's unlikely anyone will hit this, but yeah, it's technically a breaking change. I'll make it a major version bump anyway just to be safe.

**Comment by liamappelbe** [comment]
> Please add a test for this, so that we don't regress this.

Done

**Pull request merged**